openmc/lattice.py

```diff
@@ -0,0 +1,1901 @@
+from abc import ABCMeta
+from collections import OrderedDict
+from collections.abc import Iterable
+from copy import deepcopy
+from math import sqrt, floor
+from numbers import Real, Integral
+from xml.etree import ElementTree as ET
+
+import numpy as np
+
+import openmc.checkvalue as cv
+import openmc
+from openmc._xml import get_text
+from openmc.mixin import IDManagerMixin
+
+
+class Lattice(IDManagerMixin, metaclass=ABCMeta):
+    """A repeating structure wherein each element is a universe.
+
+    Parameters
+    ----------
+    lattice_id : int, optional
+        Unique identifier for the lattice. If not specified, an identifier will
+        automatically be assigned.
+    name : str, optional
+        Name of the lattice. If not specified, the name is the empty string.
+
+    Attributes
+    ----------
+    id : int
+        Unique identifier for the lattice
+    name : str
+        Name of the lattice
+    pitch : Iterable of float
+        Pitch of the lattice in each direction in cm
+    outer : openmc.Universe
+        A universe to fill all space outside the lattice
+    universes : Iterable of Iterable of openmc.Universe
+        A two-or three-dimensional list/array of universes filling each element
+        of the lattice
+
+    """
+
+    next_id = 1
+    used_ids = openmc.Universe.used_ids
+
+    def __init__(self, lattice_id=None, name=''):
+        # Initialize Lattice class attributes
+        self.id = lattice_id
+        self.name = name
+        self._pitch = None
+        self._outer = None
+        self._universes = None
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def pitch(self):
+        return self._pitch
+
+    @property
+    def outer(self):
+        return self._outer
+
+    @property
+    def universes(self):
+        return self._universes
+
+    @name.setter
+    def name(self, name):
+        if name is not None:
+            cv.check_type('lattice name', name, str)
+            self._name = name
+        else:
+            self._name = ''
+
+    @outer.setter
+    def outer(self, outer):
+        cv.check_type('outer universe', outer, openmc.Universe)
+        self._outer = outer
+
+    @staticmethod
+    def from_hdf5(group, universes):
+        """Create lattice from HDF5 group
+
+        Parameters
+        ----------
+        group : h5py.Group
+            Group in HDF5 file
+        universes : dict
+            Dictionary mapping universe IDs to instances of
+            :class:`openmc.Universe`.
+
+        Returns
+        -------
+        openmc.Lattice
+            Instance of lattice subclass
+
+        """
+        lattice_id = int(group.name.split('/')[-1].lstrip('lattice '))
+        name = group['name'][()].decode() if 'name' in group else ''
+        lattice_type = group['type'][()].decode()
+
+        if lattice_type == 'rectangular':
```

Comment:
Consider splitting the specific rectangular and hexagonal handling out into their respective classes. That will reduce the overall complexity of this classmethod.